core/src/main/java/com/datastax/oss/driver/internal/core/addresstranslation/EC2MultiRegionAddressTranslator.java

```diff
@@ -0,0 +1,151 @@
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.addresstranslation;
+
+import com.datastax.oss.driver.api.core.addresstranslation.AddressTranslator;
+import com.datastax.oss.driver.api.core.context.DriverContext;
+import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import javax.naming.Context;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.Attributes;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.InitialDirContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link AddressTranslator} implementation for a multi-region EC2 deployment <b>where clients are
+ * also deployed in EC2</b>.
+ *
+ * <p>Its distinctive feature is that it translates addresses according to the location of the
+ * Cassandra host:
+ *
+ * <ul>
+ *   <li>addresses in different EC2 regions (than the client) are unchanged;
+ *   <li>addresses in the same EC2 region are <b>translated to private IPs</b>.
+ * </ul>
+ *
+ * This optimizes network costs, because Amazon charges more for communication over public IPs.
+ *
+ * <p>Implementation note: this class performs a reverse DNS lookup of the origin address, to find
+ * the domain name of the target instance. Then it performs a forward DNS lookup of the domain name;
+ * the EC2 DNS does the private/public switch automatically based on location.
+ */
+public class EC2MultiRegionAddressTranslator implements AddressTranslator {
+
+  private static final Logger logger =
+      LoggerFactory.getLogger(EC2MultiRegionAddressTranslator.class);
+
+  private final DirContext ctx;
+
+  public EC2MultiRegionAddressTranslator(@SuppressWarnings("unused") DriverContext context) {
+    @SuppressWarnings("JdkObsolete")
+    Hashtable<Object, Object> env = new Hashtable<>();
+    env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.dns.DnsContextFactory");
+    try {
+      ctx = new InitialDirContext(env);
+    } catch (NamingException e) {
+      throw new RuntimeException("Could not create translator", e);
+    }
+  }
+
+  @VisibleForTesting
+  EC2MultiRegionAddressTranslator(DirContext ctx) {
+    this.ctx = ctx;
+  }
+
+  @Override
+  public InetSocketAddress translate(InetSocketAddress socketAddress) {
+    InetAddress address = socketAddress.getAddress();
+    try {
+      // InetAddress#getHostName() is supposed to perform a reverse DNS lookup, but for some reason
+      // it doesn't work within the same EC2 region (it returns the IP address itself).
+      // We use an alternate implementation:
+      String domainName = lookupPtrRecord(reverse(address));
+      if (domainName == null) {
+        logger.warn("Found no domain name for {}, returning it as-is", address);
```

Comment:
Could you add a log prefix? You can use `context.sessionName()` like other session-wide policies.